model/scratch/dffml_model_scratch/model/slr.py

```diff
@@ -0,0 +1,153 @@
+        usable = []
+        if len(features) != 1:
+            raise ValueError("Simple Linear Regression doesn't support features other than 1")
+        for feature in features:
+            if feature.dtype() != int and feature.dtype() != float:
+                raise ValueError("Simple Linear Regression only supports int or float feature")
+            if feature.length() != 1:
+                raise ValueError("Simple LR only supports single values (non-matrix / array)")
+            usable.append(feature.NAME)
+        return usable
+
+    async def best_fit_line(self, xs, ys):
+        m = ((mean(xs)*mean(ys)) - mean(xs*ys))/((mean(xs)*mean(xs)) - (mean(xs*xs)))
+        b = mean(ys) - (m*mean(xs))
+        regression_line = [m*x + b for x in xs]
+        accuracy = await self.coeff_of_deter(ys, regression_line)
+        await self.set_regression_line(m, b, accuracy)
+        return (m, b, accuracy)
+
+    async def predict_input(self, x):
+        prediction = self.regression_line[0]*x+self.regression_line[1]
+        return prediction
+
+    async def train(self, sources: Sources, features: Features):
+        '''
+        Train using repos as the data to learn from.
+        '''
+        feature = await self.applicable_features(features)
+        async for repo in sources.with_features(feature):
+            # Grab a subset of the feature data being stored within the repo
+            # The subset is the feature_we_care_about and the feature we are want to predict
+            feature_data = repo.features(feature + [self.parent.config.predict])
+            self.xData = np.append(self.xData, feature_data[feature[0]])
+            self.yData = np.append(self.yData, feature_data[self.parent.config.predict])
+        self.regression_line = await self.best_fit_line(self.xData, self.yData)
+        
+    async def accuracy(self, sources: Sources, features: Features) -> Accuracy:
+        '''
+        Evaluates the accuracy of our model after training using the input repos
+        as test data.
+        '''
+        if self.regression_line is None:
+            raise ValueError('Model Not Trained')
+        accuracy_value = self.regression_line[2]
+        return Accuracy(accuracy_value)
+
+    async def predict(self, repos: AsyncIterator[Repo], features: Features) -> \
+                    AsyncIterator[Tuple[Repo, Any, float]]:
+        '''
+        Uses trained data to make a prediction about the quality of a repo.
+        '''
+        feature = await self.applicable_features(features)
+        async for repo in repos:
+            feature_data = repo.features(feature)
+            yield repo, await self.predict_input(feature_data[feature[0]]), self.regression_line[2]
+
+
+@entry_point('slr')
+class SLR(Model):
+
+    CONTEXT = SLRContext
+    
+    def __init__(self, config: SLRConfig) -> None:
+        super().__init__(config)
+        self.saved = {'regression_line': None}
```

Comment:
Saved should be an empty `dict`, because the keys within it that we are assigning to will be those hash values of the features (which you havne't added yet, but it will work the same when you do if you make it an empty `dict`)

```python
self.saved = {}
```

Reply:
but features won't include regression_line? How will I save that? (I haven't used hashlib before so a bit confused)

Reply:
So we load / save from a filename based on the feature name we are predicting on. Within SLRContext the `train`, `accuracy`, and `predict` methods create and user regression lines which are based on the feature(s) they were told to use. Therefore, within `self.saved` we want to create a mapping of `features -> regression_line_applicable_to_these_features`.

By using hashlib we know we're always going to get a nice string of hex which we'll have no trouble saving and loading with regards to character encodings.

So we want to take the features that we are training, predicting, accuracy accessing on and save and load from `self.saved` using the hash of those feature names (the one feature in this case). The hash of a value is always going to be the same everytime you hash the same input value, so this is how we'll store the regression lines in saved, by the unique hash associated with the feature the line was trained to predict.

Does that makes sense? Also, check out the documentation for hashlib, play around with it and then you'll understand better how we're using it. https://docs.python.org/3/library/hashlib.html

Essentially, when you call `update` on the hash object you are having it hash more input data. you then call `digest/hexdigest` to get the resulting hash. Imagine you had a large file, and you wanted to calculate a unique value which maps only to those specific contents of that file, change the contents, new hash. You could do this by reading it line by line and passing each line to digest.